controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -512,7 +477,17 @@ public void listScopes(final SecurityContext securityContext, final AsyncRespons
                          .thenApply(scopesList -> {
                              ScopesList scopes = new ScopesList();
                              scopesList.forEach(scope -> {
-                                     scopes.addScopesItem(new ScopeProperty().scopeName(scope));
+                                 try {
+                                     if (restAuthHelper.isAuthorized(authHeader,
+                                             AuthResourceRepresentation.ofAScope(scope),
+                                             principal, READ)) {
+                                         scopes.addScopesItem(new ScopeProperty().scopeName(scope));
+                                     }
+                                 } catch (AuthException e) {
```

Comment:
Let's not use exceptions as flow control. This is not only ugly, it about 3 orders of magnitude slower. Is there a way to simply check if the principal has access?

Reply:
To clarify, we are not using the exception for the purpose of flow control here. I introduced a new method ``isAuthorized(..)`` precisely to avoid having to use exception as flow control. Under normal circumstances the exception is not expected to be encountered during authorization. 

But, ``RESTAuthHelper's isAuthorized(..)`` delegates authorization decision internally to an existing method: ``PravegaAuthManager's authorize()``, which returns true/false usually. In one specific scenario, it returns an exception. The catch block in the highlighted code in there is to handle that.